File: site/src/commonTest/kotlin/net/numa08/niconico_advertiser_list2/util/UserIdExtractorTest.kt
package net.numa08.niconico_advertiser_list2.util

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

/**
 * UserIdExtractorのテスト
 */
class UserIdExtractorTest {
    @Test
    fun testExtractUserId_directId() {
        // 直接ID
        assertEquals("753685", UserIdExtractor.extractUserId("753685"))
        assertEquals("123456", UserIdExtractor.extractUserId("123456"))
    }

    @Test
    fun testExtractUserId_userPageUrl() {
        // ユーザーページURL
        assertEquals(
            "753685",
            UserIdExtractor.extractUserId("https://www.nicovideo.jp/user/753685"),
        )
        assertEquals(
            "123456",
            UserIdExtractor.extractUserId("https://www.nicovideo.jp/user/123456"),
        )
    }

    @Test
    fun testExtractUserId_userVideoUrl() {
        // ユーザー動画ページURL
        assertEquals(
            "753685",
            UserIdExtractor.extractUserId("https://www.nicovideo.jp/user/753685/video"),
        )
        assertEquals(
            "123456",
            UserIdExtractor.extractUserId("https://www.nicovideo.jp/user/123456/video"),
        )
    }

    @Test
    fun testExtractUserId_rssUrl() {
        // RSS URL
        assertEquals(
            "753685",
            UserIdExtractor.extractUserId("https://www.nicovideo.jp/user/753685/video?rss=atom"),
        )
        assertEquals(
            "123456",
            UserIdExtractor.extractUserId("https://www.nicovideo.jp/user/123456/video?rss=atom&page=1"),
        )
    }

    @Test
    fun testExtractUserId_withFragment() {
        // フラグメント付きURL
        assertEquals(
            "753685",
            UserIdExtractor.extractUserId("https://www.nicovideo.jp/user/753685#tab"),
        )
    }

    @Test
    fun testExtractUserId_httpScheme() {
        // HTTPスキーム
        assertEquals(
            "753685",
            UserIdExtractor.extractUserId("http://www.nicovideo.jp/user/753685"),
        )
    }

    @Test
    fun testExtractUserId_subdomain() {
        // サブドメイン
        assertEquals(
            "753685",
            UserIdExtractor.extractUserId("https://sp.nicovideo.jp/user/753685"),
        )
    }

    @Test
    fun testExtractUserId_withWhitespace() {
        // 前後の空白
        assertEquals("753685", UserIdExtractor.extractUserId("  753685  "))
        assertEquals(
            "753685",
            UserIdExtractor.extractUserId("  https://www.nicovideo.jp/user/753685  "),
        )
    }

    @Test
    fun testExtractUserId_invalid() {
        // 無効な入力
        assertNull(UserIdExtractor.extractUserId("abc"))
        assertNull(UserIdExtractor.extractUserId("sm12345678")) // 動画IDではない
        assertNull(UserIdExtractor.extractUserId(""))
        assertNull(UserIdExtractor.extractUserId("   "))
    }

    @Test
    fun testExtractUserId_invalidUrl() {
        // 無効なURL
        assertNull(UserIdExtractor.extractUserId("https://example.com/user/753685"))
        assertNull(UserIdExtractor.extractUserId("https://www.nicovideo.jp/watch/sm12345678"))
        assertNull(UserIdExtractor.extractUserId("https://www.nicovideo.jp/"))
        assertNull(UserIdExtractor.extractUserId("not a url"))
    }

    @Test
    fun testExtractUserId_alphanumericUserId() {
        // 英数字混合は無効（ユーザーIDは数字のみ）
        assertNull(UserIdExtractor.extractUserId("abc123"))
        assertNull(UserIdExtractor.extractUserId("https://www.nicovideo.jp/user/abc123"))
    }
}

File: frontend/test/userIdExtractor.test.ts
// 対応仕様: docs/FRONTEND_REQUIREMENTS.md
//   FE-130〜FE-134（投稿者ID抽出）
// テストケースは現行KotlinのUserIdExtractorTest（受け入れ基準）を移植したもの
import { describe, expect, it } from "vitest";
import { extractUserId } from "../src/lib/userIdExtractor";

describe("extractUserId", () => {
  it("FE-132: 直接の投稿者IDはそのまま返す", () => {
    expect(extractUserId("753685")).toBe("753685");
    expect(extractUserId("123456")).toBe("123456");
  });

  it("FE-130: 前後の空白を除去してから判定する", () => {
    expect(extractUserId("  753685  ")).toBe("753685");
  });

  it("FE-133: ユーザーページURLから投稿者IDを抽出する", () => {
    expect(extractUserId("https://www.nicovideo.jp/user/753685")).toBe("753685");
    expect(extractUserId("http://www.nicovideo.jp/user/753685")).toBe("753685");
    expect(extractUserId("https://sp.nicovideo.jp/user/753685")).toBe("753685");
  });

  it("FE-133: /user/{id}/video や RSS URL・フラグメント付きも受理する", () => {
    expect(extractUserId("https://www.nicovideo.jp/user/753685/video")).toBe("753685");
    expect(extractUserId("https://www.nicovideo.jp/user/753685/video?rss=atom")).toBe("753685");
    expect(extractUserId("https://www.nicovideo.jp/user/753685#tab")).toBe("753685");
  });

  it("FE-131/FE-134: 数字のみでない直接入力は失敗する", () => {
    expect(extractUserId("abc")).toBeNull();
    expect(extractUserId("sm12345678")).toBeNull();
  });

  it("FE-134: 空・空白のみは失敗する", () => {
    expect(extractUserId("")).toBeNull();
    expect(extractUserId("   ")).toBeNull();
  });

  it("FE-134: 別ドメイン・/user/以外のパス・英数字混在IDは失敗する", () => {
    expect(extractUserId("https://example.com/user/753685")).toBeNull();
    expect(extractUserId("https://www.nicovideo.jp/watch/sm12345678")).toBeNull();
    expect(extractUserId("https://www.nicovideo.jp/")).toBeNull();
    expect(extractUserId("https://www.nicovideo.jp/user/abc123")).toBeNull();
  });
});
